### ISSUE TITLE
Bug with SUBSTR and CockroachDB thinking that a `INT` is a `STR` according to the logs

### ISSUE THREAD

**Issue by dvaerum** [cockroachdb/cockroach#94897]
I am testing out Nextcloud with CockroachDB. And I think I may have encountered a bug in CockroachDB, but with the database module for PostgreSQL from PHP-PDO.

**Describe the problem**

It looks like CockroachDB converts/misinterprets an integer value for a string value

**To Reproduce**

To trigger the bug, just a database (in this example 'nextcloud') and connect to it and run the following command.
```sql
CREATE TABLE IF NOT EXISTS filecache (fileid INT8 NOT NULL, storage INT8 NOT NULL DEFAULT 0, path VARCHAR(4000) NULL DEFAULT NULL, path_hash VARCHAR(32) NOT NULL DEFAULT '');

INSERT INTO filecache (fileid, storage, path, path_hash) VALUES (1, 828084885788491777, 'files/New folder/auto-dark-mode/.git', b2f888b162c19ade85a80fe0928876d7);
```

Now execute this PHP script up against your CockroachDB
```php
<?php

// Connect to the database
$host = 'db';
$dbname = 'nextcloud';
$username = 'nextcloud';
$password = 'NEXTCLOUD_PASSWORD';

$conn = null;
while ($conn === null) {
    try {
        $conn = new PDO("pgsql:host=$host;dbname=$dbname", $username, $password);
        $conn->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
        $conn->setAttribute(PDO::ATTR_AUTOCOMMIT, FALSE);
    } catch(PDOException $e) {
        echo 'Error: ' . $e->getMessage();
        sleep(2);
    }
}

sleep(3);


// Print database before Test
$sql = "SELECT * FROM filecache";
$stmt = $conn->prepare($sql);
try {
    $stmt->execute();
    $result = $stmt->fetchAll();

    // Print the result
    print("\nPrint database before Test: ");
    print_r($result);
} catch(Driver\Exception $e) {
    print_r($e);
}


// Send TEST the request
// Send TEST the request

// Bad SQL request
$sql = 'UPDATE "filecache" SET "storage" = ?, "path_hash" = MD5((CAST(? AS TEXT) || CAST(SUBSTR("path", ?) AS TEXT))), "path" = (CAST(? AS TEXT) || CAST(SUBSTR("path", ?) AS TEXT)) WHERE ("storage" = ?) AND ("path" LIKE ?)';
// Good SQL request
// $sql = 'UPDATE "filecache" SET "storage" = ?, "path_hash" = MD5((CAST(? AS TEXT) || CAST(SUBSTR("path", CAST(? AS INT)) AS TEXT))), "path" = (CAST(? AS TEXT) || CAST(SUBSTR("path", CAST(? as INT)) AS TEXT)) WHERE ("storage" = ?) AND ("path" LIKE ?)';

$stmt = $conn->prepare($sql);

#                Index  Value                                              Type (1 = int, 2 = string)
$stmt->bindValue(1    , 828084885788491777,                                1);
$stmt->bindValue(2    , "files_trashbin/files/auto-dark-mode.d1673181254", 2);
$stmt->bindValue(3    , 32,                                                1);
$stmt->bindValue(4    , "files_trashbin/files/auto-dark-mode.d1673181254", 2);
$stmt->bindValue(5    , 32,                                                1);
$stmt->bindValue(6    , 828084885788491777,                                1);
$stmt->bindValue(7    , "files/New folder/auto-dark-mode/%",               2);

try {
    $stmt->execute();
    print("\nIt works!\n");
} catch(Driver\Exception $e) {
    print_r($e);
}


// Print database after Test
$sql = "SELECT * FROM filecache";
$stmt = $conn->prepare($sql);
try {
    $stmt->execute();
    $result = $stmt->fetchAll();

    // Print the result
    print("\nPrint database after Test: ");
    print_r($result);
} catch(Driver\Exception $e) {
    print_r($e);
}

// Close the connection
$conn = null;

?>
```

and it will return the following error (even through the same code works fine on PostgreSQL)
```
Fatal error: Uncaught PDOException: SQLSTATE[23502]: Not null violation: 7 ERROR:  null value in column "path_hash" violates not-null constraint in /var/www/html/test-db.php:52
```

So at first, it just looks like a violation of a constraint, and it technically is, but there is more too it.
If we comment-out the "Bad SQL request" (line 43) and remove the comment from the "Good SQL request" (line 45) and try again, everything works.

So, what is the difference? Well, basically both places there
```sql
CAST(SUBSTR("path", ?) AS TEXT)
```
is used, it was changed to
```sql
CAST(SUBSTR("path", CAST(? AS INT)) AS TEXT))
```
Simply put, I make sure that the values declared in line 52 & 54 are converted to `INT`, even through in the request send to the database, they were already set to the type `INT`.

The problem with the value `32` in line 52 & 54 being considered a string by CockroachDB is also confirmed by looking in the log file `cockroach-data/logs/cockroach-sql-exec.log`.
```
I230108 16:38:08.241467 1653 9@util/log/event_log.go:32 ⋮ [n1,client=192.168.96.3:57802,user=‹nextcloud›] 277 ={
    "Timestamp":1673195888238600832,
    "EventType":"query_execute",
    "Statement":"UPDATE ‹\"\"›.‹\"\"›.‹filecache› SET ‹storage› = $1, ‹path_hash› = md5((CAST($2 AS STRING) || CAST(substr(‹path›, $3) AS STRING))), ‹path› = (CAST($4 AS STRING) || CAST(substr(‹path›, $5) AS STRING)) WHERE (‹storage› = $6) AND (‹path› LIKE $7)",
    "Tag":"UPDATE",
    "User":"‹nextcloud›",
    "PlaceholderValues":[
        "‹828084885788491777›",
        "‹'files_trashbin/files/auto-dark-mode.d1673181254'›",
        "‹'32'›",
        "‹'files_trashbin/files/auto-dark-mode.d1673181254'›",
        "‹'32'›",
        "‹828084885788491777›",
        "‹'files/New folder/auto-dark-mode/%'›"
    ],
    "ExecMode":"exec",
    "SQLSTATE":"23502",
    "ErrorText":"null value in column ‹\"path_hash\"› violates not-null constraint",
    "Age":2.736035,
    "FullTableScan":true,
    "TxnCounter":5
}
```
Here we can see in the `PlaceholderValues` the value `'32'` is wrapped with `'` just like the strings `'files_trashbin/files/auto-dark-mode.d1673181254'` & `'files/New folder/auto-dark-mode/%'`, but the integer `828084885788491777` is not wrapped with `'` because it is recognized as an integer.

**Expected behavior**
I expect that the function `substr(input: varbit, start_pos: int) → varbit` described in the documented:
https://www.cockroachlabs.com/docs/stable/functions-and-operators.html#string-and-byte-functions

would inform the part of CockroachDB which interprets incoming values that the value used for `start_pos` should be interpreted as an integer. Or at the very least throw an exception saying
```
The parameter `start_pos` in the function `substr` requires an integer value not an string value
```
or something like that 😁

**Additional data / screenshots**
I have submitted all my logs and SQL queries in the reproduce section

**Environment:**
 - CockroachDB version: v22.2.1 (at the time latest container image)
 - Container Image: nextcloud:latest
 - Client app: PHP-PDO

**Additional context**

I have created a docker-compose environment which reproduces the error, using my script. I hope this will help the developer(s).
The environment can switch between using CockroachDB & PostgreSQL.
[cockroach_bug.zip](https://github.com/cockroachdb/cockroach/files/10369104/cockroach_bug.zip)

Jira issue: CRDB-23214

**Comment by rafiss** [comment]
Thanks for the report. We will investigate this further soon.

**Comment by rafiss** [comment]
The reason this is happening is because `substr(<string>, <string>)` is _also_ an allowed form of the `substr` function. See the docs at https://www.cockroachlabs.com/docs/stable/functions-and-operators.html and search for `substr(input: string, regex: string) → string`. The behavior is that the function returns NULL if there is no matching substring, which explains the error you see.

This is a bit of a divergence from Postgres as you pointed out. Here is the Postgres behavior:
```
# This is what you tried, and Postgres converts the '3' to an integer.
rafiss@127:postgres> select substr('my_string', '3')
+---------+
| substr  |
|---------|
| _string |
+---------+
SELECT 1

# Spelling out `substring` causes it to try a regex match.
rafiss@127:postgres> select substring('my_string', '3')
+-----------+
| substring |
|-----------|
| <null>    |
+-----------+
SELECT 1

# Another form of the `substring` function.
rafiss@127:postgres> select substring('my_string' from '3')
+-----------+
| substring |
|-----------|
| <null>    |
+-----------+
SELECT 1

```

Addtionally, Postgres 15 contains the `regexp_substring` function, which behaves similarly:
```
regexp_substr ( string text, pattern text [, start integer [, N integer [, flags text [, subexpr integer ] ] ] ] ) → text

  Returns the substring within string that matches the N'th occurrence of the POSIX regular expression pattern, or NULL if there is no such match; see Section 9.7.3.

  regexp_substr('ABCDEF', 'c(.)(..)', 1, 1, 'i') → CDEF
  regexp_substr('ABCDEF', 'c(.)(..)', 1, 1, 'i', 2) → EF
```

So the main difference is that in CockroachDB, `substr` and `substring` behave identically, but in Postgres they are different. Also, the CockroachDB `substring` function behaves the same as the Postgres `substring` function.

Hopefully this is enough information for you to workaround the issue.

**Comment by rafiss** [comment]
Based on the above, I'm going to close this. Let me know if you're unable to find a workaround, or if this difference from Postgres is a blocker for you.

**Comment by dvaerum** [comment]
Hey @rafiss, thanks for you feedback, I however is still confused.

As I understand it CockroachDB is suppose to be compatible with PostgreSQL, but this is a incompatibility with PostgreSQL, so does this not make it a bug?

Or is there some type of system setting which can be used to switch the behavior more to be like PostgreSQL? 
Just like with ‘sql.defaults.serial_normalization.

**Comment by rafiss** [comment]
We can leave this issue open for tracking purposes, since you're right that we'd consider it a bug to be different from the Postgres behavior. 

However, we can't prioritize a fix right away, since fixing it would mean that there would be a backwards incompatible change for any users who are already using the existing implementation of `substring`. We will need to determine how much of an impact that could have.

The impact of fixing the difference and causing a breaking change may be more than the effort it takes to work around the issue. In this case the workaround should be feasible: pass in an integer argument instead of a string argument. Could you say more about why using an integer is a blocker?

**Comment by dvaerum** [comment]
I want to use CockroachDB as a drop-in replacement for PostgresSQL in my Nextcloud setup, but as long as SUBSTR doesn't work as expected, I cannot. The example code I used is triggered every time you delete files in Nextcloud.

This is the first time I am trying to replace PostgresSQL with CockroachDB, so I don't know if I will run in to problems with other project because I have not tried.

I had plans to try using it at work with some systems there we need HA and is using PostgresSQL (if I could get it working with Nextcloud), but if there are more of these cases, and they cannot be fixed/workaround either by changing the code "to the correct PostgreSQL behaviour" or by having a setting for how these "function" should work, I'm imaging something like this `sql.defaults.functions.substr = CockroachDB/PostgresSQL`.

I know that there can be a problem with changing the function `SUBSTR` if many of your customers are using the function and expect the behaviour of `SUBSTRING`, and that is why I suggest a way for people to config how the function work with a system setting. If I can flip a switch to get more PostgresSQL compatibility, that would work for me.

I hope this answers your question @rafiss about why this is a blocker for me, in short if there is not a way to make this work, I cannot use CockroachDB for Nextcloud and I would be uncomfortable with using it other places if I found another case like `SUBSTR` and there was no way forward.

**Comment by dikshant** [comment]
While PG Compatibility is something that is constantly improving, there are always edge cases where CockroachDB may not be the same. While we may tackle this in the future we don't have the cycles to work on this right now.
Compatibility with third party tools often requires extensive testing and in some cases bespoke integration. If things change on NextCloud's side in a future release things may still break. Hence, CockroachDB as a backend support would be a better request to raise at NextCloud.

**Comment by dvaerum** [comment]
Hey @dikshant 

> While PG Compatibility is something that is constantly improving, there are always edge cases where CockroachDB may not be the same. While we may tackle this in the future we don't have the cycles to work on this right now.

Thanks for the info, but then CockroachDB is not what I am looking for, maybe I will check back in after 2–3 years

> Compatibility with third party tools often requires extensive testing and in some cases bespoke integration. If things change on NextCloud's side in a future release things may still break. Hence, CockroachDB as a backend support would be a better request to raise at NextCloud.

Yes, that was my exact concern and why I wanted to see how compatibility bugs like this one were handled.
As things are, right now. I am sadly getting the feeling that they are not a high priority for you guys, or you don't have a good way to handle them.
Which is sadly not gonna work for me.

**Comment by banholzer** [comment]
I have the same concern like @dvaerum.

I think something like a switch to change behavior to more "psql like" would be really great. As long as cockroachdb does not show they are able to handle this kind of a request this project disqualifies itself for professional use in areas where you want to replace psql with a cloud-native db.

**Comment by dikshant** [comment]
We recently discussed a policy to gate these changes behind a cluster setting that enables the Postgres compatible behavior. Our plan is to make the new behavior the default after a release and remove the cluster setting in the future.

We are starting with another issue where we have similarly diverged in behavior and will tackle this when we have some bandwidth.

https://github.com/cockroachdb/cockroach/issues/104371

**Comment by remram44** [comment]
There is now a long list of incompatibilities in CockroachDB's docs: [Features that differ from PostgreSQL](https://www.cockroachlabs.com/docs/v23.2/postgresql-compatibility#features-that-differ-from-postgresql) Maybe this issue can be closed?

I just tested NextCloud again and it doesn't get far into the installation process before crashing. It would make more sense for NextCloud to consider supporting CockroachDB than the other way around.

**Comment by mcassaniti** [comment]
> There is now a long list of incompatibilities in CockroachDB's docs: [Features that differ from PostgreSQL](https://www.cockroachlabs.com/docs/v23.2/postgresql-compatibility#features-that-differ-from-postgresql) Maybe this issue can be closed?

That would go against [this comment above](https://github.com/cockroachdb/cockroach/issues/94897#issuecomment-1396344348) "... since you're right that we'd consider it a bug to be different from the Postgres behavior". Would you mind @rafiss providing any feedback? I didn't see this specific difference of features documented on the compatibility page.

Having said that, has anyone on this issue reached out to the Nextcloud developers and asked about considering support for CockroachDB? Perhaps it would be a win-win to showcase how much change to an existing large codebase is required to support CockroachDB if PostgreSQL support exists.

**Comment by dvaerum** [comment]
> Having said that, has anyone on this issue reached out to the Nextcloud developers and asked about considering support for CockroachDB? Perhaps it would be a win-win to showcase how much change to an existing large codebase is required to support CockroachDB if PostgreSQL support exists.

Nope, I was considering it, because I only had to enable on feature one CockroachDB and change on line in Nextcloud, and it was up and running. Don’t know if it was fully compatible, because I never ran the Nextcloud database unittests. However, I lost interest, because I have gotten the feeling that CockroachDB have little focus on ensuring compatibility with PostgreSQL and since I wanted to use it with other products too that is just too big a concern for me.

**Comment by dvaerum** [comment]
> It would make more sense for NextCloud to consider supporting CockroachDB than the other way around.

I disagree, since this issue/bug is about Cockroach being incompatible with PostgreSQL, which is one of there selling points. They them self have acknowledged that this is a bug.
Nextcloud was just the tool just used for testing

**Comment by remram44** [comment]
I assumed the situation had changed since the ticket was opened, since there are even more issues with the current version of NextCloud (didn't install for me) and there is now a long list of incompatibilities on the CockroachDB docs. The docs make it clear that apps require porting from PostgreSQL to CockroachDB, so I wouldn't expect it to work based on that. If I should, that's great, but maybe the docs need to be updated to reflect that goal?
